src/main/java/io/lettuce/core/protocol/CommandHandler.java

```diff
@@ -561,9 +561,7 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf buffer) throws Interrup
                 }
             }
 
-            if (buffer.refCnt() != 0) {
-                buffer.discardReadBytes();
```

Comment:
How about `ByteBuf.discardSomeReadBytes()`? Wouldn't that have the same effect benefiting from the actual `ByteBuf` implementation?

Reply:
I considered that, but according to docs "(...) this method might discard some, all, or none of read bytes depending on its internal implementation". The actual implementation in `AbstractByteBuf` does indeed exactly what I did, but there is no guarantee. So I thought it's better to do it on our own for a better control.

Reply:
Looking in `ByteToMessageDecoder`, `ByteBuf. discardSomeReadBytes()` is what netty does for the majority of message decoders. There's also an additional optimization that they release a buffer if it's no longer needed.

In Lettuce 4.x, we shared a `CommandHandler` across multiple `Channel`s which made it impossible to tweak the `buffer`. We could actually apply a similar optimization. The buffer is allocated in `channelRegistered`. What if we release the `buffer` if it does not contain any readable bytes anymore? That would require retaining the buffer for quite a while but I guess following a similar path to what `ByteToMessageDecoder` does seems quite reasonable to me. 

WDYT?

Btw. Lettuce does not use `ByteToMessageDecoder` because we started with our own decoder/state machine and there was no compelling reason to switch to a different decoding scheme.

Reply:
If we use `discardSomeReadBytes`, I'd like to still have the guarantee that we don't call it to often - e.g. `ByteToMessageDecoder` calls it after (by default) 16 successful reads - WDYT?

How about instead of releasing, just shrinking the buffer if there are no bytes to read? There is a `capacity(int newCapacity)` method, we can use it if the buffer is not readable and it has grown to size larger than, let's say, 4x initial capacity?

Reply:
I'm okay with the reduction after some time iff reads did not yield a command completion. IMO we could also read directly from the received buffer without writing the data to the intermediate buffer and only write the remaining bytes to the intermediate buffer. This could speed-up responses for small commands but introduces another level of complexity to maintain properly the intermediate buffer.

Reply:
I would like this fix to be merged ASAP, so let me suggest the following: I change the call of `discardReadBytes` to `discardSomeReadBytes` as you sugguested in the beginning, and then open another issue and/or PR, where we can discuss how to use the buffers in the most optimal way?

Reply:
I think `discardSomeReadBytes(…)` should be fine for the beginning.